front/components/assistant_builder/AssistantBuilderTablesModal.tsx

```diff
@@ -55,45 +114,199 @@ export default function AssistantBuilderTablesModal({
   const [selectedDataSource, setSelectedDataSource] =
     useState<DataSourceType | null>(null);
 
-  const [selectedTable, setSelectedTable] =
-    useState<AssistantBuilderTableConfiguration | null>(null);
+  const [selectedManagedTables, setSelectedManagedTables] = useState<
+    ContentNode[]
+  >([]);
+  const [isSaving, setIsSaving] = useState(false);
+
+  const [isInitializingNodes, setIsInitializingNodes] = useState(false);
+  const [nodeIntializationError, setNodeIntializationError] =
+    useState<boolean>(false);
+
+  const [parentsById, setParentsById] = useState<Record<string, Set<string>>>(
+    {}
+  );
+  const [isInitializingParents, setIsInitializingParents] = useState(false);
+  const [parentsIntializationError, setParentsIntializationError] =
+    useState<boolean>(false);
+
+  const initializeManagedTables = useCallback(
+    async (
+      dataSource: DataSourceType,
+      tableConfigs: AssistantBuilderTableConfiguration[]
+    ) => {
+      const tableIds = tableConfigs.map((config) => config.tableId);
+
+      const contentNodes = await getContentNodes(
+        owner.sId,
+        dataSource.name,
+        tableIds
+      );
+
+      if (contentNodes.length !== tableIds.length) {
+        throw new Error(
+          `Failed to fetch content nodes for all tables. Expected ${tableIds.length}, got ${contentNodes.length}.`
+        );
+      }
+
+      setSelectedManagedTables((prev) => [
+        ...prev.filter(
+          (n) => !contentNodes.some((c) => c.internalId === n.internalId)
+        ),
+        ...contentNodes,
+      ]);
+    },
+    [owner.sId, setSelectedManagedTables]
+  );
+
+  const initializeParents = useCallback(
+    async (internalIds: string[], dataSource: DataSourceType) => {
+      const parents = await getParents(owner.sId, dataSource.name, [
+        ...internalIds,
+      ]);
+      setParentsById(parents);
+    },
+    [owner.sId]
+  );
+
+  useEffect(() => {
+    // When coming back to the modal, if the user selects a data source that
+    // is a connection and there are already tables selected, we need to
+    // fetch the content nodes for those tables and add them to the selected
+    // managed tables.
+
+    if (
+      !selectedDataSource ||
+      !selectedDataSource.connectorId ||
+      isInitializingNodes ||
+      !!selectedManagedTables.length ||
+      !Object.keys(tablesQueryConfiguration).length ||
+      !!nodeIntializationError
+    ) {
```

Comment:
Can we make this a bit more explicit, there are a lot of `!` 😅 
```suggestion
    if (
      !selectedDataSource ||
      !selectedDataSource.connectorId ||
      isInitializingNodes ||
      selectedManagedTables.length  > 0||
      Object.keys(tablesQueryConfiguration).length === 0 ||
      !!nodeIntializationError
    ) {
```

Reply:
I don't really feel your suggestion makes it a lot more explicit ? I know you dislike relying on falsy-ness of `.length` but personally I prefer this versus manually comparing with 0. TBH I see it as a big failure from JS that an empty array is not falsy in itself (i.e, I should be able to just do `! selectedManagedTables`)

Reply:
Agree, probably splitting this into variables could make it easier to parse?

Reply:
Did that !